src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,831 @@ const std::string& TNAuthList::Entry::telephone_number() const {
+
+template <typename T>
+std::optional<std::vector<T>> sort_and_merge_ranges(std::optional<std::span<const T>> ranges) {
+   // This method iteratively both sorts and combines IPAddressOrRange or ASIdOrRange objects.
+   // cf. https://www.rfc-editor.org/rfc/rfc3779.html#section-2.2.3.6 and https://www.rfc-editor.org/rfc/rfc3779.html#section-3.2.3.4
+   // This implementation uses only min-max ranges internally, so sorting by the prefix length is not necessary / impossible here.
+   //
+   // We first sort all range objects by their minimum value, then check if two adjacent elements can be combined into one
+   // (either one ends and the other immediately starts, or they overlap), in which case the two are removed, and a new combined object is added.
+   // This process is repeated until there are no further changes, because multiple adjacent elements may need to be combinded one by one.
+
+   if(!ranges.has_value()) {
+      return std::nullopt;
+   }
+
+   std::vector<T> current(ranges.value().begin(), ranges.value().end());
+
+   if(current.empty()) {
+      return current;
+   }
+   bool changed = true;
+   while(changed) {
+      changed = false;
+      // sort by size of the min value
+      std::sort(current.begin(), current.end(), [](T& a, T& b) { return a.min() < b.min(); });
+      for(size_t i = 0; i < current.size() - 1; i++) {
+         const T a = current[i];
+         const T b = current[i + 1];
+         // they are either adjacent or overlap
+         if(b.min() <= a.max() || b.min() == (a.max() + 1)) {
+            // erase old a and b
+            current.erase(current.begin() + i, current.begin() + i + 2);
+            current.insert(current.begin() + i, T(a.min(), std::max(a.max(), b.max())));
+            // look again at the new element we just created, might need to be merged with others still
+            i -= 1;
+            changed = true;
+         }
+      }
+   }
+   return current;
+}
+
+template <typename T>
+bool validate_subject_in_issuer(const std::vector<T>& subject, const std::vector<T>& issuer) {
+   // ensures that the subject ranges are enclosed by the issuer ranges
+   // both vectors are already sorted, so we can do this in O(n+m)
+
+   // the issuer has 0 ranges to validate against, so this can only work if the subject also has none
+   if(issuer.size() == 0) {
+      return subject.size() == 0;
+   }
+   for(auto subj = subject.begin(), issu = issuer.begin(); subj != subject.end();) {
+      // the subject is outside of the closest issuer range on the left (min) side
+      if(subj->min() < issu->min()) {
+         return false;
+      }
+      // the issuer range is smaller than the subject range, step to the next issuer range
+      if(subj->min() > issu->max()) {
+         issu++;
+         // we have run out of issuer ranges, but still have subject ranges left to validate
+         if(issu == issuer.end() && subj != subject.end()) {
+            return false;
+         }
+         continue;
```

Comment:
The `continue` here makes the logic harder to parse. (At least for me :sweat_smile:) Can this be done some other way? IIUC each of these are exclusive so could be done as `if else`